### ISSUE TITLE
not-callable false positive when calling a "cached callable"

### ISSUE THREAD

**Issue by mhooreman** [PyCQA/pylint#2900]
### Steps to reproduce
Create a class having a _f property and a f "property decorated" method which sets _f to a callable and returns it.
```python
import math

class C():
    def __init__(self):
        self._f = None
    @property
    def f(self):
        if self._f is None:
            self._f = math.sin
        return self._f

if __name__ == "__main__":
    print(C().f(0))  # <====== Issue comes here
```

### Current behavior
```not-callable(E1102): false_not_callable.py:14:10: : C().f is not callable```

### Expected behavior
No issue

### pylint --version output
```
pylint 2.3.1
astroid 2.2.5
Python 3.7.3 (default, Mar 27 2019, 09:40:28) 
[GCC 7.3.0]
```

**Comment by PCManticore** [comment]
Thanks for the report @mhooreman This is a classic example of pylint not understanding flow control, it does not understand the reassignment of `_f` to an actual callable function under the `if` guard.
